File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    n = len(diffs)

    # 특정 level로 퍼즐을 모두 풀 수 있는지 확인하는 함수
    def can_clear(level):
        total = 0

        # 첫 퍼즐은 이전 퍼즐이 없으므로 무조건 times[0]
        total += times[0]
        if total > limit:
            return False

        for i in range(1, n):
            diff = diffs[i]
            time_cur = times[i]
            time_prev = times[i-1]

            if level >= diff:
                total += time_cur
            else:
                fail = diff - level
                total += fail * (time_cur + time_prev) + time_cur

            if total > limit:
                return False

        return True

    # level은 1 ~ max(diffs) 사이에서 이분 탐색
    left, right = 1, max(diffs)
    answer = right

    while left <= right:
        mid = (left + right) // 2

        if can_clear(mid):
            answer = mid
            right = mid - 1
        else:
            left = mid + 1

    return answer

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.swift
import Foundation

func solution(_ diffs: [Int], _ times: [Int], _ limit: Int64) -> Int {
    let n = diffs.count
    let maxDiff = diffs.max() ?? 1

    func can(_ level: Int) -> Bool {
        var total: Int64 = 0

        // 0번 퍼즐
        total += Int64(times[0])
        if total > limit { return false }

        if n == 1 { return total <= limit }

        for i in 1..<n {
            let diff = diffs[i]
            let tCur = Int64(times[i])
            if diff <= level {
                total += tCur
            } else {
                let x = Int64(diff - level)
                let tPrev = Int64(times[i - 1])
                total += (tCur + tPrev) * x + tCur
            }
            if total > limit { return false } // 오버 방지 + 조기 종료
        }
        return total <= limit
    }

    var lo = 1
    var hi = maxDiff  // 이 정도면 무조건 가능(틀림 0) -> 문제에서 항상 해답 존재

    while lo < hi {
        let mid = (lo + hi) / 2
        if can(mid) {
            hi = mid
        } else {
            lo = mid + 1
        }
    }
    return lo
}